Fetch error handling

jakewan opened the pull request:
This PR adds a hook that providers can use when a problem occurs making a measurement. Examples include DNS misconfiguration, firewall issues, and non-success HTTP status responses.

This PR is to update the `main-with-cbsi-patches` branch, which Doppler.js should target when built for production. In the future, a separate branch will be created, omitting the CBSi-only patches. This will be used to submit a PR to the openinsights upstream repo.

jakewan committed:
Omit ClientSettings import/export from entry module

It's still unclear why, but the ClientSettings import/export here causes
a problem in `parcel build` with the --experimental-scope-hoisting flag.

Some relevant Github issues:

* https://github.com/parcel-bundler/parcel/issues/2566

* https://github.com/parcel-bundler/parcel/issues/1540

jakewan committed:
Update README.md [cbsi-patch]

jakewan committed:
Add bare CBSI-PATCHES.md [cbsi-patch]

jakewan committed:
Comments about possible error conditions

jakewan committed:
Add provider hook for error reporting

Includes:

* HTTP status code validation in Fetch.test

* Update docs

jakewan committed:
Merge branch 'main-with-cbsi-patches' into fetch-error-handling

jakewan committed:
Add TestSetupResult arg to onTestFailure callback

jakewan committed:
Update docs

jakewan commented:
@vnsparacio I'm working on the broken unit tests now.

jakewan committed:
Fix broken unit tests

jakewan committed:
Add support for aborting slow fetch requests after user-defined timeout

jakewan committed:
Mark fetch tests that trigger timeout for use later by provider

jakewan committed:
Make timeoutTriggered property optional

Seemed to be causing problems with unit tests run by Github Actions.

vnsparacio commented on CBSI-PATCHES.md:
```diff
@@ -0,0 +1,3 @@
+# CBSi Patches
+
+TODO
```
Can we just delete this file until we're ready to put it in?

jakewan replied:
I'd rather leave it as a placeholder.

jakewan committed:
Let provider specify the beacon data type

jakewan committed:
Remove HTTP status code validation

Instead expect providers to look at the response status when they have
the opportunity to do so in createFetchTestResult and possibly
downstream in makeBeaconData.

jakewan committed:
Extend the abstraction model to support more provider customization.

jakewan committed:
Support beacon custom headers

jakewan committed:
Support for specifying headers; use Fetch API POST
